### ISSUE TITLE
Compound launch is quite confusing, especially compare to Multi-target debugging

### ISSUE THREAD

**Issue by qiulang** [Microsoft/vscode-recipes#102]
I found compound launch is quite confusing, especially compare to multi-target debugging (with multi-root workspace).

First of all, the document said [An alternative way to start multiple debug session...](https://code.visualstudio.com/docs/editor/debugging#_compound-launch-configurations) so why choose one over the other? What is the advantage Compound launch has that Multi-target hasn't ?

Second, it is hard to find a useful example/recipe that really works! I read the article [simultaneous “nirvana” ](https://medium.com/@auchenberg/introducing-simultaneous-nirvana-javascript-debugging-for-node-js-and-chrome-in-vs-code-d898a4011ab1) and the timey example https://github.com/auchenberg/timey/tree/master/.vscode. But if I need to do the step 3 "Start your node server with debugging enabled by running". then I really prefer to just using multi-target debugging. Because with multi-target debugging there is no extra typing, no extra editing to launch.json. 

BTW, should I add "compounds" setting to client's `launch.json` or server's `launch.json` or it doesn't matter? And how do I make sure compounds launch run the tasks serially, i.e. start server first then launch Chrome ?

Third, I need to start webpack server to proxy the request to server (seems quite common these days), i.e. there are 3 steps in my debugging process and I just can't make it work with Compound launch.

Before I use multi-root workspace, my debugging process goes like these, start a vs-code for server side codes, launch the debugger. Then start another vs-code for client side codes, launch webpack server, then launch chrome (using Chrome Debugger).

With multi-root workspace, the process is the same but with one vs-code. So first choose to start server debugger, then start webpack in terminal then launch chrome.

I tried to, but server task and webpack task simultaneously.  

```
...
"compounds": [
        {
            "name": "Server/Client",
            "configurations": ["NPM server","webpack","vuejs: chrome"]
        }
    ]
```

**Comment by auchenberg** [comment]
Lots of questions here @qiulang, let me try to unfold:

1) Compounds are a way for you to run multiple debug configurations with the click of a button. It's a shortcut, so you don't have to start your multiple debugging targets manually.

2) If you need to start a webpack server or something else, you shouldn't create a new debugging configuration for this, but use a `npm script` or use a use the built-in task system in VS Code: https://code.visualstudio.com/docs/editor/tasks

3) You can choose to use compounds in any of your workspace folders's `launch.json` and make a reference between. See https://github.com/Microsoft/vscode/issues/38134#issuecomment-360187394

4) If you have a specific sequence of things that needs to happen, my advice would be to use an npm script to run things. See https://stackoverflow.com/questions/39172536/running-npm-scripts-sequentially. A compound will launch both debug configurations in parallel.

For your flow it sounds like having a `npm start` script for your back-end, and a  `npm start` for your front-end that runs your webpack dev server, and your vuejs scripts would be the way to go. A compound debug configuration would then start your backend, and your front-end.

**Comment by auchenberg** [comment]
Let me know if there's further questions, and I can re-open this issue, if there's any action items on our side.

**Comment by qiulang** [comment]
Hi, I suddenly realize there is an important point in your answer, please correct me if I am wrong: so for compound launch the server part, it is to **attach** to my server NOT **launch** my server, right?

When I use multi-target debugging I use "launch via NPM" to start my server, then a "npm start" to run webpack server (in terminal), then launch chrome. But according to your suggestion, to use compound launch, I should then run a npm script in terminal to launch my server and webpack server(sequentially), then use the compound launch to attach the debugger to my server and launch chrome. Is that correct ?

**Comment by qiulang** [comment]
I feel your advice to use a npm script to start my server then webpack server (if I understand you correctly) may not work. 

Because when I start my server it won't exit and the next script won't be fire! Just like poststart hook won't fire when the start script is to start a node server. 

I even opened an issue for npm for that today :$  https://github.com/npm/npm/issues/20651

**Comment by auchenberg** [comment]
You can choose to either have the VS Code debuggers attaching to existing processes, or use launch configurations to launch stuff for you. The point here is that you want to have ONE NPM task that set things up for you, and then have VS Code to use that NPM task.  VS Code is in a way "dumb" as it only does what you tell it to do, so if your NPM script pauses on a particular step, and never manages to start your process with debugging enabled, then our debugger can't connect.

To me it sounds like your workflow are like:

1. Start node http server (to run your api, etc)
2. Start webpack dev server (to compile and serve your front-end)
4. Attach Node.js debugger to your Node http server
3. Launch Chrome for debugging, that visits localhost:3000 (or similar)

For 1), 2) you might want to look into https://www.npmjs.com/package/concurrently

And for your VS Code debug configuration https://github.com/auchenberg/timey/blob/master/.vscode/launch.json should do it, if you for example choose to run "npm start" from your integrated terminal followed by starting debugging in VS Code.

**Comment by qiulang** [comment]
No my workflow does not need step no. 3, only 1,2,4. Actually the step 3 was my question that I raised to you regarding the compound launch (after you closed the issue).  It has been several days so let me restate it.

Because starting webpack dev server will make things complicated, let me first leave it out.

So I was trying to use Compound launch to <1> Start node http server from vscode and also attach debugger to it <2> Launch Chrome for debugging. 

With multi-target debugging it is easy, just 2 clicks. But I was not able to do that using compound launch. Then I read your article "simultaneous nirvana” and your example, found that you actually use compound launch to  **attached** "Node.js debugger to the Node http server", **which means I need to start my node server before compound launch** (in your article).  If that is the case, then of course I will prefer multi-target debugging over compound launch. 

I was not sure if I used compound launch open I raised the issue to you.

Using webpack dev server make things even more complicated. I actually tried https://www.npmjs.com/package/concurrently before.  But I didn't make it work. 

So far I am still using multi-root workspace, multi-target debugging process and my workflow is this, 

1.  one click to start server and debugger
2. in the terminal, start webpack dev server
3. another click to launch chrome for debugging.
